tests/flytekit/unit/core/test_type_engine.py

```diff
@@ -2499,16 +2497,29 @@ class DatumMashumaro(DataClassJSONMixin):
 
     @dataclass_json
     @dataclass
-    class Datum(DataClassJSONMixin):
+    class DatumDataclassJson(DataClassJSONMixin):
         x: int
         y: Color
 
-    transformer = DataclassTransformer()
+    @dataclass
+    class DatumDataclass:
+        x: int
+        y: Color
```

Comment:
Can you add another test to see what happens with `typing.Union`? Specifically:

```python
@dataclass
class DatumDataUnion:
    path: typing.Union[str, os.PathLike]
```

If the `path` started out as a `pathlib.Path(...)`, does it deserialize into a string or a `Path` object?

Reply:
I found that TypeTransformer can't handle support for typing.Union[str, os.PathLike] or typing.Union[str, FlyteFile].
(I thought that `os.PathLike` is decided not supported here.
https://docs.flyte.org/en/latest/api/flytekit/generated/flytekit.types.file.FlyteFile.html#flytekit.types.file.FlyteFile.path

Support `typing.Union[str, FlyteFile]` in flyte could be an enhancement.
For more details, I need to trace more code to find the core reason why we can support these cases now, but it definitely is not because of this PR.

Reply:
Hmm, thank you for checking. When I suggested `typing.Union[str, os.PathLike]`, I was thinking about `FlyteFile`. But `FlyteFile` has it's own type transformer, so it's okay.

In any case, I'm okay with the current scope of this PR.